### PR TITLE
Implementation of authentication, session creation and messaging

### PR DESCRIPTION
**Pull request by vishal-1408**
# Pull Request Type
Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

# Purpose
- To implement authentication, session creation and messaging

# Does this PR introduce any new bug?
- [ ] Yes
- [x] No
- [ ] Maybe

### CODE REVIEW

File: db/index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+exports.updateOne = (connection, config) => {
+  return new Promise((resolve, reject) => {
+    connection.query(`UPDATE ${config.tables} SET ${config.fields} WHERE ${config.conditions}`, config.values, (error, result) => {
+      if (error) reject(error)
+      resolve(result)
+    })
+  })
+}
```
Comment (mdhishaamakhtar) [inline db/index.js]: SQL Injection Vulnerability!

Comment (vishal-1408) [reply db/index.js]: Okay! I will use type ORM and push the commits to this PR

File: api/controllers/chat.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      where: {
+        sentIn: sId
+      }
+    })
+
+    const oldMessages = []
+    let sender = null
+
+    for (let i = 0; i <= messages.length - 1; i++) {
+      sender = await User.findOne({ where: { userId: messages[i].sentBy }, attributes: { exclude: ['password', 'emailVerified', 'otp', 'otpExpiry'] } })
```
Comment (mdhishaamakhtar) [inline api/controllers/chat.js]: Can be done in one query instead of multiple

File: api/controllers/task.js [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    const tasks = await Task.findAll({
+      where: {
+        givenIn: sessionId,
+        assignedTo: req.user.userId
+      }
+    })
+
+    for (let i = 0; i <= tasks.length - 1; i++) {
+      tasks[i].assignedTo = await tasks[i].getUser({ attributes: { exclude: ['password', 'otp', 'otpExpiry', 'emailVerified'] } })
+      tasks[i].createdBy = await User.findOne({ where: { userId: tasks[i].createdBy }, attributes: { exclude: ['password', 'otp', 'otpExpiry', 'emailVerified'] } })
```
Comment (mdhishaamakhtar) [inline api/controllers/task.js]: Can be done in one query again

File: api/controllers/task.js [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        givenIn: sessionId
+      }
+    })
+
+    console.log(tasks)
+
+    for (let i = 0; i <= tasks.length - 1; i++) {
+      tasks[i].assignedTo = await tasks[i].getUser({ attributes: { exclude: ['password', 'otp', 'otpExpiry', 'emailVerified'] } })
+      tasks[i].createdBy = await User.findOne({ where: { userId: tasks[i].createdBy }, attributes: { exclude: ['password', 'otp', 'otpExpiry', 'emailVerified'] } })
+    }
```
Comment (mdhishaamakhtar) [inline api/controllers/task.js]: One query

**Review by mdhishaamakhtar — CHANGES_REQUESTED** [review]
There are several places where you can avoid calling the database inside a loop. Do some array operations and all inside the loop, call the db once, then map accordingly. If you call db inside a loop and say on the 7th iteration for some reason your app crashes, those first 6 iterations would have already written the data while the ones after that will not. And rollback well :) so you might as well just avoid something like this. Its a mistake I've made in the past and a mistake I deeply regret making. Just a suggestion
Apart from that everything looks great! Good job!

File: db/index.js (hunk 1 above)
Comment (mdhishaamakhtar) [reply db/index.js]: Okay you can resolve this now @vishal-1408

File: api/controllers/chat.js (hunk 2 above)
Comment (vishal-1408) [reply api/controllers/chat.js]: Yeah I will update it and keep this in mind

**Review by SaurusXI — APPROVED** [review]

**Pull request merged**